### PR TITLE
fully switch to @solana/buffer-layout

### PR DESCRIPTION
**Pull request by israelidanny**
So, here is some more maintenance work of migrating off the non-typescript `buffer-layout`.

A couple of observations:
- It's quite nerve-wrecking to contribute here because there are no tests 😅  How do we make sure even a harmless update such as this one doesn't ruin stuff?
- The lending program JS code heavily depended on generics in the buffer-layout typings, so instead of depending on the npm version I depended on https://github.com/jordansexton/buffer-layout/tree/generics and made sure it builds without errors. I would consider merging https://github.com/jordansexton/buffer-layout/tree/generics (or at least discussing why not merge it)

### CODE REVIEW

**Comment by jordansexton** [comment]
Nice, I'll check this out! This might be useful too: https://github.com/solana-labs/buffer-layout-utils/

My idea was to replace some of the stuff in the lending buffers with this, but it depends on the generics branch too. There's also an open PR to implement the token program bindings in TypeScript cc @jstarry

**Comment by jordansexton** [comment]
@israelidanny generics branch has been merged, `@solana/buffer-layout` has been released on NPM (v4.0.0)

**Comment by jordansexton** [comment]
`@solana/buffer-layout-utils` has also been released (v0.1.0) using it.

**Comment by israelidanny** [comment]
@jordansexton I updated the branch to use the generics version of buffer-layout (4.0.0) and the utils.

**Review by jordansexton — DISMISSED** [review]
Looks great! Just had a question about the change from `Buffer` -> `Uint8Array` in a few places.

**Comment by jordansexton** [comment]
@israelidanny it looks like the token lending JS tests failed, could you merge from master into here and see if it fixes the issue?

**Comment by bcrant** [comment]
Hey, I went ahead and merged latest master into @israelidanny 's changes to keep the ball rolling: https://github.com/solana-labs/solana-program-library/pull/2779

cc: @jordansexton

**Comment by israelidanny** [comment]
@bcrant I went ahead and did the same here. This PR is now up to date with master.
@jordansexton wen merge?

File: token-lending/js/src/instructions/initLendingMarket.ts
Comment (israelidanny) [reply token-lending/js/src/instructions/initLendingMarket.ts:11]: 1. Uint8Array is more portable (Buffer is a node specific implementation of Uint8Array)
2. Caused compilation errors with TypeScript so I had to remove

**Comment by jordansexton** [comment]
I just commented on the other PR, but the tests failed. Will plan to merge when the tests pass!

File: token-lending/js/src/instructions/initLendingMarket.ts
Comment (jordansexton) [reply token-lending/js/src/instructions/initLendingMarket.ts:11]: Yeah, I was originally thinking it made sense to just use Buffer since we already have it from buffer-layout, but that lib bundles it, and we don't need to create buffers here anyway.

**Comment by jordansexton** [comment]
Can you also upgrade buffer-layout-utils to v0.2.0? I just published it, it has some fixes for ESM that @steveluscher helped with.

**Comment by israelidanny** [comment]
@jordansexton I'll do it now :)

**Comment by israelidanny** [comment]
@jordansexton there were some issues with the buffer-utils themselves and web3 and TypeScript, so I enabled "skipLibCheck" to prevent the building failing on the account of `node_modules` issues. 

```
$ rm -rf lib/* && tsc && rollup -c && rm lib/rollup.config.d.ts
node_modules/@solana/buffer-layout-utils/node_modules/@solana/web3.js/lib/index.d.ts:2:1 - error TS6200: Definitions of the following identifiers conflict with those in another file: Struct, Enum, SOLANA_SCHEMA, MAX_SEED_LENGTH, PublicKeyInitData, PublicKeyData, PublicKey, Account, Blockhash, BPF_LOADER_DEPRECATED_PROGRAM_ID, EpochSchedule, NONCE_ACCOUNT_LENGTH, NonceAccount, Keypair, MessageHeader, CompiledInstruction, MessageArgs, Message, TransactionSignature, PACKET_DATA_SIZE, AccountMeta, TransactionInstructionCtorFields, SerializeConfig, TransactionInstruction, SignaturePubkeyPair, TransactionCtorFields, NonceInformation, Transaction, TokenAccountsFilter, Context, SendOptions, ConfirmOptions, ConfirmedSignaturesForAddress2Options, SignaturesForAddressOptions, RpcResponseAndContext, Commitment, Finality, LargestAccountsFilter, GetLargestAccountsConfig, GetSupplyConfig, SignatureStatusConfig, ContactInfo, VoteAccountInfo, VoteAccountStatus, InflationGovernor, InflationReward, EpochInfo, LeaderSchedule, Version, SimulatedTransactionAccountInfo, SimulatedTransactionResponse, ParsedInnerInstruction, TokenBalance, ParsedConfirmedTransactionMeta, CompiledInnerInstruction, ConfirmedTransactionMeta, TransactionResponse, ConfirmedTransaction, PartiallyDecodedInstruction, ParsedMessageAccount, ParsedInstruction, ParsedMessage, ParsedTransaction, ParsedConfirmedTransaction, BlockResponse, ConfirmedBlock, ConfirmedBlockSignatures, PerfSample, Supply, TokenAmount, TokenAccountBalancePair, AccountBalancePair, SlotUpdate, SlotInfo, ParsedAccountData, StakeActivationData, DataSlice, MemcmpFilter, DataSizeFilter, GetProgramAccountsFilter, GetProgramAccountsConfig, GetParsedProgramAccountsConfig, AccountInfo, KeyedAccountInfo, AccountChangeCallback, ProgramAccountChangeCallback, SlotChangeCallback, SlotUpdateCallback, SignatureResultCallback, SignatureStatusNotification, SignatureReceivedNotification, SignatureSubscriptionCallback, SignatureSubscriptionOptions, RootChangeCallback, Logs, LogsFilter, LogsCallback, SignatureResult, TransactionError, TransactionConfirmationStatus, SignatureStatus, ConfirmedSignatureInfo, HttpHeaders, FetchMiddleware, ConnectionConfig, Connection, BPF_LOADER_PROGRAM_ID, BpfLoader, CreateEd25519InstructionWithPublicKeyParams, CreateEd25519InstructionWithPrivateKeyParams, Ed25519Program, Loader, STAKE_CONFIG_ID, Authorized, Lockup, CreateStakeAccountParams, CreateStakeAccountWithSeedParams, InitializeStakeParams, DelegateStakeParams, AuthorizeStakeParams, AuthorizeWithSeedStakeParams, SplitStakeParams, WithdrawStakeParams, DeactivateStakeParams, MergeStakeParams, StakeInstruction, StakeInstructionType, StakeAuthorizationType, StakeAuthorizationLayout, StakeProgram, CreateAccountParams, TransferParams, AssignParams, CreateAccountWithSeedParams, CreateNonceAccountParams, CreateNonceAccountWithSeedParams, InitializeNonceParams, AdvanceNonceParams, WithdrawNonceParams, AuthorizeNonceParams, AllocateParams, AllocateWithSeedParams, AssignWithSeedParams, TransferWithSeedParams, SystemInstruction, SystemInstructionType, SystemProgram, CreateSecp256k1InstructionWithPublicKeyParams, CreateSecp256k1InstructionWithEthAddressParams, CreateSecp256k1InstructionWithPrivateKeyParams, Secp256k1Program, VALIDATOR_INFO_KEY, Info, ValidatorInfo, VOTE_PROGRAM_ID, Lockout, EpochCredits, VoteAccount, SYSVAR_CLOCK_PUBKEY, SYSVAR_INSTRUCTIONS_PUBKEY, SYSVAR_RECENT_BLOCKHASHES_PUBKEY, SYSVAR_RENT_PUBKEY, SYSVAR_REWARDS_PUBKEY, SYSVAR_STAKE_HISTORY_PUBKEY, SendTransactionError, Cluster, LAMPORTS_PER_SOL

2 declare module '@solana/web3.js' {
  ~~~~~~~

  node_modules/@solana/web3.js/lib/index.d.ts:2:1
    2 declare module '@solana/web3.js' {
      ~~~~~~~
    Conflicts are in this file.

node_modules/@solana/web3.js/lib/index.d.ts:2:1 - error TS6200: Definitions of the following identifiers conflict with those in another file: Struct, Enum, SOLANA_SCHEMA, MAX_SEED_LENGTH, PublicKeyInitData, PublicKeyData, PublicKey, Account, Blockhash, BPF_LOADER_DEPRECATED_PROGRAM_ID, EpochSchedule, NONCE_ACCOUNT_LENGTH, NonceAccount, Keypair, MessageHeader, CompiledInstruction, MessageArgs, Message, TransactionSignature, PACKET_DATA_SIZE, AccountMeta, TransactionInstructionCtorFields, SerializeConfig, TransactionInstruction, SignaturePubkeyPair, TransactionCtorFields, NonceInformation, Transaction, TokenAccountsFilter, Context, SendOptions, ConfirmOptions, ConfirmedSignaturesForAddress2Options, SignaturesForAddressOptions, RpcResponseAndContext, Commitment, Finality, LargestAccountsFilter, GetLargestAccountsConfig, GetSupplyConfig, SignatureStatusConfig, ContactInfo, VoteAccountInfo, VoteAccountStatus, InflationGovernor, InflationReward, EpochInfo, LeaderSchedule, Version, SimulatedTransactionAccountInfo, SimulatedTransactionResponse, ParsedInnerInstruction, TokenBalance, ParsedConfirmedTransactionMeta, CompiledInnerInstruction, ConfirmedTransactionMeta, TransactionResponse, ConfirmedTransaction, PartiallyDecodedInstruction, ParsedMessageAccount, ParsedInstruction, ParsedMessage, ParsedTransaction, ParsedConfirmedTransaction, BlockResponse, ConfirmedBlock, ConfirmedBlockSignatures, PerfSample, Supply, TokenAmount, TokenAccountBalancePair, AccountBalancePair, SlotUpdate, SlotInfo, ParsedAccountData, StakeActivationData, DataSlice, MemcmpFilter, DataSizeFilter, GetProgramAccountsFilter, GetProgramAccountsConfig, GetParsedProgramAccountsConfig, AccountInfo, KeyedAccountInfo, AccountChangeCallback, ProgramAccountChangeCallback, SlotChangeCallback, SlotUpdateCallback, SignatureResultCallback, SignatureStatusNotification, SignatureReceivedNotification, SignatureSubscriptionCallback, SignatureSubscriptionOptions, RootChangeCallback, Logs, LogsFilter, LogsCallback, SignatureResult, TransactionError, TransactionConfirmationStatus, SignatureStatus, ConfirmedSignatureInfo, HttpHeaders, FetchMiddleware, ConnectionConfig, Connection, BPF_LOADER_PROGRAM_ID, BpfLoader, CreateEd25519InstructionWithPublicKeyParams, CreateEd25519InstructionWithPrivateKeyParams, Ed25519Program, Loader, STAKE_CONFIG_ID, Authorized, Lockup, CreateStakeAccountParams, CreateStakeAccountWithSeedParams, InitializeStakeParams, DelegateStakeParams, AuthorizeStakeParams, AuthorizeWithSeedStakeParams, SplitStakeParams, WithdrawStakeParams, DeactivateStakeParams, MergeStakeParams, StakeInstruction, StakeInstructionType, StakeAuthorizationType, StakeAuthorizationLayout, StakeProgram, CreateAccountParams, TransferParams, AssignParams, CreateAccountWithSeedParams, CreateNonceAccountParams, CreateNonceAccountWithSeedParams, InitializeNonceParams, AdvanceNonceParams, WithdrawNonceParams, AuthorizeNonceParams, AllocateParams, AllocateWithSeedParams, AssignWithSeedParams, TransferWithSeedParams, SystemInstruction, SystemInstructionType, SystemProgram, CreateSecp256k1InstructionWithPublicKeyParams, CreateSecp256k1InstructionWithEthAddressParams, CreateSecp256k1InstructionWithPrivateKeyParams, Secp256k1Program, VALIDATOR_INFO_KEY, Info, ValidatorInfo, VOTE_PROGRAM_ID, Lockout, EpochCredits, VoteAccount, SYSVAR_CLOCK_PUBKEY, SYSVAR_INSTRUCTIONS_PUBKEY, SYSVAR_RECENT_BLOCKHASHES_PUBKEY, SYSVAR_RENT_PUBKEY, SYSVAR_REWARDS_PUBKEY, SYSVAR_STAKE_HISTORY_PUBKEY, SendTransactionError, Cluster, LAMPORTS_PER_SOL

2 declare module '@solana/web3.js' {
  ~~~~~~~

  node_modules/@solana/buffer-layout-utils/node_modules/@solana/web3.js/lib/index.d.ts:2:1
    2 declare module '@solana/web3.js' {
      ~~~~~~~
    Conflicts are in this file.


Found 2 errors.
```

**Comment by israelidanny** [comment]
Merged upstream again, so fully caught up as of now.

**Comment by bcrant** [comment]
🙌  Thank you! @israelidanny

**Comment by jordansexton** [comment]
@israelidanny can you give me write access on your fork? It helps if you open a PR from a branch so we can make quick updates / resolve conflicts before merging

I'll rerun the tests now

**Comment by jordansexton** [comment]
Hmm, I can't rerun the tests, it appears there are merge conflicts in it.

**Comment by israelidanny** [comment]
Sure let me give you full rights to expedite it

**Review by jac18281828 — APPROVED** [review]
Thanks for making this change!   I could use this as well!  🙏

**Comment by israelidanny** [comment]
@jordansexton I sent you an invite to my branch

**Comment by jordansexton** [comment]
Hmm, I'm seeing `This branch cannot be rebased due to conflicts`... but I can't find any conflicts. Any idea what's going on here?

**Comment by jordansexton** [comment]
Oh, I see. It was an issue from the "rebase and merge" option. Nevermind!

**Comment by jordansexton** [comment]
I have some fixes here that haven't been pushed. Trying to find time to get to it this week, haven't forgotten about it.

**Comment by israelidanny** [comment]
@jordansexton alrighty! I am keeping an eye on this issue if I'm needed in any way or form. 
We're also keenly awaiting for this to drop over at https://github.com/metaplex-foundation/js/issues/11

**Pull request merged**
